### PR TITLE
global: Invenio-Circulation Entities

### PR DESCRIPTION
**Pull request by mvesper**
### There are the following transitions currently:

```
                                  lose_item
+------------------+-------------------+----------+----------------------------------+
|                  |                   |          |                                  |
|                  |                   v          |                                  |
|                  |   +------------+MISSING      |                                  |
|                  |   |                          |                                  |
|                  |   |return_missing_item       |                                  |
|                  |   |                          |                                  |
|                  +   v       loan_item          +                                  |
|               ON_SHELF+------------------------>ON_LOAN<------------+              |
|               +   ^  ^                          +   +               |              |
|               |   |  |                          |   |               |              |
|               |   |  +--------------------------+   |               |              |
|               |   |         return_item         |   |               |              |
|               |   |                             |   |               |              |
|               |   |                             |   |overdue_item   |extend_loan   |
|   process_item|   |return_processed_item        |   |               |              |
|               |   |                             |   |               |              |
|               |   |                             |   |               |              |
|               v   +                             +   v               |              |
+--------------+IN_PROCESS                        OVERDUE+------------+              |
                                                        +                            |
                                                        +----------------------------+
```
### How do we handle [multiple] requests?

Currently there is the array field __circulation.holdings_
Do we want to encode implicit information about the type of that hold (loan/request) in the _start_date_ attribute and the position? (If _start_date_ is older/equal to today, and the holds is at position 0 -> active loan) (mind _import this_ though)

### CODE REVIEW

File: invenio_circulation/api.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+            raise PIDInvalidAction()
+
+        return method(self, *args, **kwargs)
+    return wrapper
+
+
+class CirculationStatus(object):
+    """Class holding all availabe circulation item statuses."""
+
+    ON_SHELF = u'on_shelf'
```
Comment (jirikuncar) [inline invenio_circulation/api.py]: please do not use `u` prefix here

File: invenio_circulation/api.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    def wrapper(self, *args, **kwargs):
+        """Check current deposit status."""
+        if self['_circulation']['status'] not in statuses:
+            raise PIDInvalidAction()
+
+        return method(self, *args, **kwargs)
+    return wrapper
+
+
+class CirculationStatus(object):
```
Comment (jirikuncar) [inline invenio_circulation/api.py]: can you move the status class to `models.py`?

Comment (jirikuncar) [inline invenio_circulation/api.py]: What about calling it `ItemStatus`?

File: invenio_circulation/api.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    """Class holding all availabe circulation item statuses."""
+
+    ON_SHELF = u'on_shelf'
+    ON_LOAN = u'on_loan'
+    MISSING = u'missing'
+    IN_PROCESS = u'in_process'
+    OVERDUE = u'overdue'
```
Comment (jirikuncar) [inline invenio_circulation/api.py]: IMHO `overdue` is not status but a calculated property of an item object.

File: invenio_circulation/api.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        """
+        for i, holding in enumerate(self['_circulation']['holdings'][:]):
+            if holding['id_'] == id_:
+                del self['_circulation']['holdings'][i]
+                return
+        else:
+            raise Exception('Holding id does not exist.')
+
+    @check_status(statuses=[CirculationStatus.ON_LOAN,
+                            CirculationStatus.OVERDUE])
```
Comment (jirikuncar) [inline invenio_circulation/api.py]: One should be able to extend loan even when it's not overdue ...

File: invenio_circulation/schemas/circulation/item.json [hunk 5]
```diff
@@ -0,0 +1,34 @@
+{
```
Comment (jirikuncar) [inline invenio_circulation/schemas/circulation/item.json]: please add version suffix to all schemas `*-v1.0.0.json`

File: invenio_circulation/schemas/circulation/loan_cycle.json [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+        "user": {"type": "object"},
+        "delivery": {"enum": ["delivery", "pickup"]},
+        "group_uuid": {"type": ["string", "null"]},
+        "start_date": {"type": "string",
+                       "format": "date-time"},
+        "end_date": {"type": "string",
+                     "format": "date-time"},
+        "desired_start_date": {"type": "string",
+                               "format": "date-time"},
+        "desired_end_date": {"type": "string",
```
Comment (jirikuncar) [inline invenio_circulation/schemas/circulation/loan_cycle.json]: can you reformat all JSON files (e.g. https://jsonformatter.curiousconcept.com/)

**Comment by jirikuncar** [comment]
I'm not sure about `OVERDUE` and `IN_PROCESS` states. Can you remove them for now to make the PR simpler and easier to validate?

**Comment by mvesper** [comment]
### Reduced transitions:

```
               lose_item
   +-------------------+----------+
   |                   |          |
   |                   v          |
   |   +------------+MISSING      |
   |   |                          |
   |   |return_missing_item       |
   |   |                          |
   +   v       loan_item          +
ON_SHELF+------------------------>ON_LOAN
       ^                          +
       |                          |
       +--------------------------+
              return_item
```

File: invenio_circulation/api.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        data = data or {}
+        schema = current_app.config.get('CIRCULATION_LOCATION_SCHEMA', None)
+
+        if schema:
+            ptu = current_app.extensions['invenio-jsonschemas'].path_to_url
+            data.setdefault('$schema', ptu(schema))
+        return super(CirculationLocation, cls).create(data, id_=id_)
+
+
+class CirculationItem(Record):
```
Comment (jirikuncar) [inline invenio_circulation/api.py]: You already have a namespace so you don't need `Circulation` prefix for class names.

File: invenio_circulation/api.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    def wrapper(self, *args, **kwargs):
+        """Check current deposit status."""
+        if self['_circulation']['status'] not in statuses:
+            raise PIDInvalidAction()
+
+        return method(self, *args, **kwargs)
+    return wrapper
+
+
+class CirculationLocation(Record):
```
Comment (jirikuncar) [inline invenio_circulation/api.py]: see https://github.com/inveniosoftware/invenio-circulation/pull/12/files#r68752409

**Comment by jirikuncar** [comment]
@mvesper we can merge it when you strip the `Circulation` prefix from class names (when you are done please remove WIP).

File: invenio_circulation/api.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+    @classmethod
+    def create(cls, data, id_=None):
+        """Create a location instance and store it in database."""
+        data = data or {}
+        schema = current_app.config.get('CIRCULATION_ITEM_SCHEMA', None)
+
+        if schema:
+            ptu = current_app.extensions['invenio-jsonschemas'].path_to_url
+            data.setdefault('$schema', ptu(schema))
+        if '_circulation' not in data:
```
Comment (jirikuncar) [inline invenio_circulation/api.py:81]: `data.setdefault('_circulation', {...})`

File: invenio_circulation/schemas/circulation/location-v1.0.0.json [hunk 10]
```diff
@@ -0,0 +1,19 @@
+{
+   "type":"object",
+   "title":"Location",
+   "properties":{
```
Comment (jirikuncar) [inline invenio_circulation/schemas/circulation/location-v1.0.0.json:4]: can we point to DoJSON holding schema?

File: invenio_circulation/schemas/circulation/item-v1.0.0.json [hunk 11]
```diff
@@ -0,0 +1,47 @@
+{
+   "type":"object",
+   "title":"Item",
+   "properties":{
+      "control_number":{
```
Comment (jirikuncar) [inline invenio_circulation/schemas/circulation/item-v1.0.0.json:5]: https://github.com/inveniosoftware/invenio-circulation/pull/12/files#r68769408

**Pull request merged**